src/collapse/perf.rs

```diff
@@ -82,35 +92,59 @@ pub struct Folder {
     /// Function entries on the stack in this entry thus far.
     stack: VecDeque<String>,
 
-    opt: Options,
+    // Options...
+    annotate_jit: bool,
+    annotate_kernel: bool,
+    include_addrs: bool,
+    include_pid: bool,
+    include_tid: bool,
+    nthreads: usize,
+}
+
+impl From<Options> for Folder {
+    fn from(mut opt: Options) -> Self {
+        opt.include_pid = opt.include_pid || opt.include_tid;
+        let event_filter_state = EventFilterState::from(opt.event_filter);
+        Self {
+            cache_line: Vec::default(),
+            event_filter_state,
+            in_event: false,
+            occurrences: Occurrences::new(opt.nthreads),
+            skip_stack: false,
+            stack: VecDeque::default(),
+            pname: String::default(),
+
+            annotate_jit: opt.annotate_jit,
+            annotate_kernel: opt.annotate_kernel,
+            include_addrs: opt.include_addrs,
+            include_pid: opt.include_pid,
+            include_tid: opt.include_tid,
+            nthreads: opt.nthreads,
+        }
+    }
+}
+
+impl Default for Folder {
+    fn default() -> Self {
+        Options::default().into()
+    }
 }
 
 impl Collapse for Folder {
-    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    fn collapse<R, W>(&mut self, reader: R, writer: W) -> io::Result<()>
     where
         R: BufRead,
         W: Write,
     {
-        let mut line = String::new();
-        loop {
-            line.clear();
+        if self.nthreads <= 1 {
+            self.collapse_single_threaded(reader)?;
+        } else {
+            self.collapse_multi_threaded(reader)?;
+        }
 
-            if reader.read_line(&mut line)? == 0 {
-                break;
-            }
+        self.occurrences.write(writer)?;
```

Comment:
Can you just drop the `?;` here and the `Ok(())` below?

Reply:
👍